Firestore/core/src/firebase/firestore/api/firestore.mm

```diff
@@ -0,0 +1,183 @@
+
+#import "FIRFirestoreSettings.h"
+#import "Firestore/Source/API/FIRCollectionReference+Internal.h"
+#import "Firestore/Source/API/FIRDocumentReference+Internal.h"
+#import "Firestore/Source/API/FIRTransaction+Internal.h"
+#import "Firestore/Source/API/FIRWriteBatch+Internal.h"
+#import "Firestore/Source/Core/FSTFirestoreClient.h"
+
+#include "Firestore/core/src/firebase/firestore/api/document_reference.h"
+#include "Firestore/core/src/firebase/firestore/auth/firebase_credentials_provider_apple.h"
+#include "Firestore/core/src/firebase/firestore/core/transaction.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/resource_path.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/executor_libdispatch.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "absl/memory/memory.h"
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+using firebase::firestore::api::Firestore;
+using firebase::firestore::auth::CredentialsProvider;
+using firebase::firestore::core::DatabaseInfo;
+using firebase::firestore::core::Transaction;
+using firebase::firestore::model::DocumentKey;
+using firebase::firestore::model::ResourcePath;
+using util::AsyncQueue;
+using util::Executor;
+using util::ExecutorLibdispatch;
+
+Firestore::Firestore(std::string project_id,
+                     std::string database,
+                     std::string persistence_key,
+                     std::unique_ptr<CredentialsProvider> credentials_provider,
+                     std::unique_ptr<AsyncQueue> worker_queue)
+    : database_id_{std::move(project_id), std::move(database)},
+      credentials_provider_{std::move(credentials_provider)},
+      persistence_key_{std::move(persistence_key)},
+      worker_queue_{std::move(worker_queue)} {
+  settings_ = [[FIRFirestoreSettings alloc] init];
+}
+
+AsyncQueue* Firestore::worker_queue() {
+  return [client_ workerQueue];
+}
+
+FIRFirestoreSettings* Firestore::settings() const {
+  std::lock_guard<std::mutex> lock{mutex_};
+  // Disallow mutation of our internal settings
+  return [settings_ copy];
+}
+
+void Firestore::set_settings(FIRFirestoreSettings* settings) {
+  std::lock_guard<std::mutex> lock{mutex_};
+  // As a special exception, don't throw if the same settings are passed
+  // repeatedly. This should make it more friendly to create a Firestore
+  // instance.
+  if (client_ && ![settings_ isEqual:settings]) {
+    HARD_FAIL(
+        "Firestore instance has already been started and its settings can "
+        "no longer be changed. You can only set settings before calling any "
+        "other methods on a Firestore instance.");
+  }
+  settings_ = [settings copy];
+}
+
+FIRCollectionReference* Firestore::GetCollection(
+    absl::string_view collection_path, FIRFirestore* firestore) {
+  EnsureClientConfigured();
```

Comment:
If it weren't for `EnsureClientConfigured`, I'd leave this method in `FIRFirestore`. Because `this` cannot be converted to `FIRFirestore`, as a workaround `FIRFirestore` is passed as a parameter, which is a temporary hack.

Reply:
The changes in #2507 make it possible to avoid this. They also allow us to keep `EnsureClientConfigured` down here.

Reply:
Done.